shadowsocks/config.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowsocks
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/eycorsican/go-tun2socks/common/log"
+)
+
+// ProxyConfig represents a Shadowsocks proxy configuration.
+type ProxyConfig struct {
+	ID         string
+	Host       string `json:"server"`
+	Port       int    `json:"server_port"`
+	Password   string
+	Cipher     string `json:"method"`
+	Name       string `json:"remarks,omitempty"`
+	Plugin     string `json:"plugin,omitempty"`
+	PluginOpts string `json:"plugin_opts,omitempty"`
+}
+
+// OnlineConfigRequest encapsulates a request to an online config server.
+type OnlineConfigRequest struct {
+	// URL is the HTTPs endpoint of an online config server.
+	URL string
+	// Method is the HTTP method to use in the request.
+	Method string
+	// TrustedCertFingerprint is the sha256 hash of the online config server's
+	// TLS certificate.
+	TrustedCertFingerprint []byte
+}
+
+// OnlineConfigResponse encapsulates a response from an online config server.
+type OnlineConfigResponse struct {
+	// OnlineConfig is the parsed server response.
+	OnlineConfig OnlineConfig
+	// HTTPStatusCode is the HTTP status code of the response.
+	HTTPStatusCode int
+	// RedirectURL is the Location header of a HTTP redirect response.
+	RedirectURL string
+}
+
+// OnlineConfig represents a SIP008 response from an online config server.
+type OnlineConfig struct {
+	Proxies []ProxyConfig `json:"servers"`
+	Version int
+}
+
+// FetchOnlineConfig retrieves Shadowsocks proxy configurations per SIP008:
+// https://github.com/shadowsocks/shadowsocks-org/wiki/SIP008-Online-Configuration-Delivery
+//
+// Pins the trusted certificate when req.TrustedCertFingerprint is non-empty.
+// Sets the response's RedirectURL when the status code is a redirect.
+// Returns an error if req.URL is a non-HTTPS URL, if there is a connection
+// error to the server, or if parsing the configuration fails.
+func FetchOnlineConfig(req OnlineConfigRequest) (*OnlineConfigResponse, error) {
```

Comment:
I would also want us to separate how we follow redirects and fetch the page from how we process the contents. Those can be defined separately and have their own nuances. Mixing them together makes it harder to understand and change.

We need some layers:
1) Fetch Access Code config: given an access code, returns a simple Outline proxy config (an Outline concept, not a SS concept) and the new access code, when applicable. This will run different logic whether it's a ss:// url or a ssconf/https.
2) Parse ss link: will simply parse the ss link and return the Outline config
3) Fetch online config
    1. given a ssconf/https url, extract the fingerprint, issue the HTTPS request (verifying the fingerprint).
    2. Parse the SS config, convert to the Outline config model and return if no redirect
    3. If redirected, save permanent redirect and call FetchAccessCode on the new url. Return that result, adding the new access code if absent from the response
4) Fetch HTTPS with fingerprint, which takes a http.Request + FP, does one simple roundtrip and returns the http response. Used by Fetch online config

I expressed as a recursion, but you can express it as a loop, so you avoid stack growth.

You also need to handle possible redirect cycles to prevent malicious access codes from DDoSing others.

Reply:
I can refactor Transport creation (cert pinning) into a separate file.

Regarding layering: my intention is for this file to implement SIP008 per the spec (read: ss concepts). I am working on implementing Outline-specific concepts in the `outline/shadowsocks/` package, where we will expose library functions.

As for redirects: why do we need to support ss:// URL redirection? The online config spec requires JSON for the final delivery. I like the idea of automatically following redirects (parsing ssconf:// URLs if needed) and returning the final URL for permanent redirects.

Also, are you suggesting to implement proxy configuration selection policy in Go? I think it may be useful to return the proxy config list to the web app so we can implement retry logic there.